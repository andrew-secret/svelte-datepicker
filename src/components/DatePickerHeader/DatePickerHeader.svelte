<script lang="ts">
  import type { IUtils } from "@date-io/core/IUtils";
  import type { Writable } from "svelte/store";
  import type { view } from "../../utils/view-types";
  import { getContext } from "svelte";

  export let dateAdapter: IUtils<Date>;
  export let currentMonth: Date;
  export let minDate: Date;
  export let maxDate: Date;
  export let selectPreviousMonth: () => void;
  export let selectNextMonth: () => void;
  export let toggleYearPicker: () => void;

  const view = getContext<Writable<view>>("view");

  $: shouldDisableMinDate = dateAdapter.isSameMonth(currentMonth, minDate);
  $: shouldDisableMaxDate = dateAdapter.isSameMonth(currentMonth, maxDate);

</script>

<div class="datepicker-header">
  <div class="year-switcher-wrapper">
    <time
      datetime={dateAdapter.format(currentMonth, "fullDate")}
      data-testid="selected-month"
      aria-live="polite"
      class="year-switcher-label"
    >
      {dateAdapter.format(currentMonth, "month")}
    </time>
    <time
      datetime={dateAdapter.format(currentMonth, "fullDate")}
      data-testid="selected-year"
      aria-live="polite"
      class="year-switcher-label"
    >
      {dateAdapter.format(currentMonth, "year")}
    </time>
    <button
      class="month-switcher"
      type="button"
      aria-pressed={$view === "years"}
      on:click={toggleYearPicker}
      data-testid="expand-button"
    >
      <span class="month-switcher-label">
        <svg
          class="caret"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
          aria-hidden="true"
        >
          <path d="M7 10l5 5 5-5z" />
        </svg>
      </span>
    </button>
  </div>
  {#if $view === "days"}
    <div class="month-switcher-wrapper">
      <button
        disabled={shouldDisableMinDate}
        class="month-switcher previous-month"
        aria-label="previous month"
        on:click={selectPreviousMonth}
      >
        <span class="month-switcher-label">
          <svg
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            aria-hidden="true"
            data-testid="ArrowLeftIcon"
          >
            <path
              d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
            />
          </svg>
        </span>
      </button>
      <button
        disabled={shouldDisableMaxDate}
        aria-label="next month"
        class="month-switcher next-month"
        on:click={selectNextMonth}
      >
        <span class="month-switcher-label">
          <svg
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            aria-hidden="true"
            data-testid="ArrowRightIcon"
          >
            <path d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z" />
          </svg>
        </span>
      </button>
    </div>
  {/if}
</div>

<style lang="scss">
  .datepicker-header {
    display: flex;
    justify-content: space-between;
    height: 50px;
    padding: 4px 0;
    box-sizing: border-box;
  }

  .year-switcher-wrapper {
    display: flex;
    align-items: center;
    padding-left: 12px;
  }

  .year-switcher-label {
    font-size: 1rem;
    margin-right: 4px;
    font-family: var(--sdp-font-family);
    font-weight: var(--sdp-font-weight);
  }

  .month-switcher-wrapper {
    display: flex;
  }

  .month-switcher {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 42px;
    height: 42px;
    border: none;
    padding: 0;
    cursor: pointer;
    background-color: transparent;
    border-radius: var(--sdp-border-radius);
    transition: background-color 200ms ease-in-out;

    &[aria-pressed]:focus {
      outline: none;
      box-shadow: 0 0 0 0.15rem var(--sdp-bg-focus-color);
    }

    &[aria-pressed="true"] {
      span {
        transform: rotate(180deg);
      }
    }

    &[aria-pressed="false"] {
      span {
        transform: rotate(0deg);
      }
    }

    &:hover {
      color: var(--sdp-hover-color);
      background-color: var(--sdp-hover-bg-color);
    }

    &:focus {
      outline: none;
      box-shadow: inset 0 0 0 0.15rem var(--sdp-bg-focus-color);
    }

    &:disabled {
      opacity: 0.35;
      pointer-events: none;
    }
  }

  .previous-month {
    margin-right: 4px;
  }

  .month-switcher:active {
    color: var(--sdp-active-color);
    background-color: var(--sdp-bg-active-color);
  }

  .month-switcher-label {
    width: 20px;
    height: 20px;
  }

</style>
