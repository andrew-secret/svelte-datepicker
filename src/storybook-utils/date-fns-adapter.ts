import LuxonAdapter from '@date-io/luxon';
import DateFnsAdapter from "@date-io/date-fns";

export const dateFnsUtils = new DateFnsAdapter();
export const luxonUtils = new LuxonAdapter();