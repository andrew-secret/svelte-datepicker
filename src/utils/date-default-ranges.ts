  export const defaultMinDate = new Date("1900-01-01");
  export const defaultMaxDate = new Date("2099-12-31");