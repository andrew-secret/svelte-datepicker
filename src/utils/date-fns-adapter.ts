import LuxonAdapter from '@date-io/luxon';
import DateFnsAdapter from "@date-io/date-fns";
import DayjsAdapter from '@date-io/dayjs'
import MomentUtils from '@date-io/moment'

// Should only be used in Storybook 
export const dateFnsUtils = new DateFnsAdapter();
export const luxonUtils = new LuxonAdapter();
export const dayJsUtils = new DayjsAdapter();
export const momentUtils = new MomentUtils();
