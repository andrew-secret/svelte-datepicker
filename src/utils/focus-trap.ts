export function handleFocusTrap(event: KeyboardEvent, {element, classNames, isFocusTrapDisabled}): void {
    if (!element || isFocusTrapDisabled) {
      return;
    }

    const focusableElements = element.querySelectorAll(classNames);
    const firstFocusableElement = focusableElements[0];
    const lastFocusableElement =
      focusableElements[focusableElements.length - 1];
    const isTabPressed = event.key === "Tab";

    if (!isTabPressed) {
      return;
    }

    /*
      the focusElements.length !== -1 check is needed to ensure that the 
      focusable elements array contains all elements. The reason for 
      this behavior is DOM manipulation due to the year picker.
    */
    if (
      !event.shiftKey &&
      document.activeElement === lastFocusableElement &&
      focusableElements.length !== 1
    ) {
      (firstFocusableElement as HTMLButtonElement).focus();
      event.preventDefault();
    }

    if (
      event.shiftKey &&
      document.activeElement === firstFocusableElement &&
      focusableElements.length !== 1
    ) {
      (lastFocusableElement as HTMLButtonElement).focus();
      event.preventDefault();
    }
  }