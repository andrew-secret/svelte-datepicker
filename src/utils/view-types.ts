export type view = "days" | "years";