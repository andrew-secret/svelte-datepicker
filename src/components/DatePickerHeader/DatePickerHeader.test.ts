import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/svelte';
import { enGB, fr, ru } from 'date-fns/locale';
import { defaultMaxDate, defaultMinDate } from '../../utils/date-default-ranges';
import { dateFnsUtils } from '../../utils/date-fns-adapter';
import DatePickerHeader from './DatePickerHeader.svelte';

describe('DatePickerHeader', () => {
    beforeEach(() => {
        dateFnsUtils.locale = enGB;
    })
    const currentMonth = new Date('2021-01-04');

    const datePickerHeaderProps = {
        dateAdapter: dateFnsUtils,
        currentMonth,
        minDate: defaultMinDate,
        maxDate: defaultMaxDate,
        selectPreviousMonth: jest.fn(),
        selectNextMonth: jest.fn(),
        toggleYearPicker: jest.fn(),
    }

    it('should render component', () => {
        const { container } = render(DatePickerHeader, {
            props: datePickerHeaderProps,
        });

        expect(container).toMatchSnapshot();
    });

    it('should show selected year and month', () => {
        const { getByTestId } = render(DatePickerHeader, {
            props: datePickerHeaderProps
        });

        const selectedYear = getByTestId('selected-year');
        const selectedMonth = getByTestId('selected-month');

        expect(selectedYear).toHaveTextContent('2021');
        expect(selectedMonth).toHaveTextContent('January');
    });

    it('should show selected month w/ french locale', () => {
        dateFnsUtils.locale = fr;
        const { getByTestId } = render(DatePickerHeader, {
            props: datePickerHeaderProps,
        });

        const selectedMonth = getByTestId('selected-month');

        expect(selectedMonth).toHaveTextContent('janvier');
    });

    it('should show selected month w/ russian locale', () => {
        dateFnsUtils.locale = ru;
        const { getByTestId } = render(DatePickerHeader, {
            props: datePickerHeaderProps,
        });

        const selectedMonth = getByTestId('selected-month');

        expect(selectedMonth).toHaveTextContent('январь');
    });

    it('should show selected month w/ english locale', () => {
        const { getByTestId } = render(DatePickerHeader, {
            props: datePickerHeaderProps
        });

        const selectedMonth = getByTestId('selected-month');

        expect(selectedMonth).toHaveTextContent('January');
    });
});
