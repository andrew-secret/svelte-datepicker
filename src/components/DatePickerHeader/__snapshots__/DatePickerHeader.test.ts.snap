// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePickerHeader should render component 1`] = `
<body>
  <div>
    <div
      class="datepicker-header svelte-18t5wxc"
    >
      <div
        class="year-switcher-wrapper svelte-18t5wxc"
      >
        <time
          aria-live="polite"
          class="year-switcher-label svelte-18t5wxc"
          data-testid="selected-month"
          datetime="4 Jan 2021"
        >
          January
        </time>
         
        <time
          aria-live="polite"
          class="year-switcher-label svelte-18t5wxc"
          data-testid="selected-year"
          datetime="4 Jan 2021"
        >
          2021
        </time>
         
        <button
          aria-pressed="false"
          class="month-switcher svelte-18t5wxc"
          data-testid="expand-button"
          type="button"
        >
          <span
            class="month-switcher-label svelte-18t5wxc"
          >
            <svg
              aria-hidden="true"
              class="caret"
              fill="currentColor"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </span>
        </button>
      </div>
       
    </div>
  </div>
</body>
`;
